src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/resource/IncentiveLevelResource.kt

```diff
@@ -86,4 +92,133 @@ class IncentiveLevelResource(
+        description = "Unauthorized to access this endpoint",
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))]
+      ),
+      ApiResponse(
+        responseCode = "403",
+        description = "Incorrect permissions to use this endpoint",
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))]
+      ),
+      ApiResponse(
+        responseCode = "404",
+        description = "Incentive level not found",
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))]
+      ),
+    ]
+  )
+  suspend fun updateIncentiveLevel(
+    @Schema(description = "Incentive level code", example = "STD", required = true, minLength = 3, maxLength = 6)
+    @PathVariable code: String,
+    @RequestBody incentiveLevel: IncentiveLevel
+  ): IncentiveLevel {
+    ensure {
+      if (code != incentiveLevel.code) {
+        errors.add("Incentive level codes must match in URL and payload")
+      }
+    }
+    return partiallyUpdateIncentiveLevel(code, incentiveLevel.toUpdate())
+  }
+
+  @PatchMapping("levels/{code}")
+  @Operation(
+    summary = "Updates an incentive level",
+    description = "Partial updates are allowed",
+    // TODO: decide and explain what happens to associated per-prison data
+    responses = [
+      ApiResponse(
+        responseCode = "200",
+        description = "Incentive level updated"
+      ),
+      ApiResponse(
+        responseCode = "400",
+        description = "Invalid payload", // TODO: maybe also when deactivating and there are prisons with level activated?
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))]
+      ),
+      ApiResponse(
+        responseCode = "401",
+        description = "Unauthorized to access this endpoint",
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))]
+      ),
+      ApiResponse(
+        responseCode = "403",
+        description = "Incorrect permissions to use this endpoint",
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))]
+      ),
+      ApiResponse(
+        responseCode = "404",
+        description = "Incentive level not found",
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))]
+      ),
+    ]
+  )
+  suspend fun partiallyUpdateIncentiveLevel(
+    @Schema(description = "Incentive level code", example = "STD", required = true, minLength = 3, maxLength = 6)
+    @PathVariable code: String,
+    @RequestBody update: IncentiveLevelUpdate
+  ): IncentiveLevel {
+    ensure {
+      update.description?.let {
+        ("description" to it).hasLengthAtLeast(1)
+      }
+    }
+    return incentiveLevelService.updateIncentiveLevel(code, update)
```

Comment:
Same comment as above, the implementation suggests maybe the code for the PUT endpoint got mixed up with the code for the PATCH endpoint (verb traditionally used for partial updates)

Also, I notice both methods get the same DTO class, how do these endpoint distinguish the two?

TBH I'm wondering if we could just have the PATCH endpoint (more flexible) - do we need the PUT endpoint too? (genuine question, I may miss something)

Reply:
The two endpoints accept different request bodies! PUT requires a complete and valid object while PATCH accepts a partial one.

On the point of do we need it: to be honest, we don't really need any of the modifying endpoints as it'll essentially never happen ;)

Reply:
I agree it's overkill, but 🤷